### PR TITLE
feat: add `assert/is-well-formed-string`

### LINKED ISSUE #1065: [RFC]: add `@stdlib/assert/is-well-formed-string`
### Description

This RFC proposes adding support for asserting whether an input value is a well-formed string (i.e. does not contain lone surrogates).

Package: `@stdlib/assert/is-well-formed-string`
Alias: `isWellFormedString`

### Related Issues

None.

### Questions

No.

### Other

Resources:

- Lone surrogates: <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String#utf-16_characters_unicode_code_points_and_grapheme_clusters>
- isWellFormed: <https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/isWellFormed>

Prior art in stdlib:

- `@stdlib/assert/is-string`
- `@stdlib/assert/is-empty-string`

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

### PR DESCRIPTION
**Pull request by gunjjoshi**
This commit adds the assert/is-well-formed-string module for checking well-formed-strings.

Resolves #1065 

## Description

Added the is-well-formed-string package. It checks for four things to check if a string is well formed or not:

  1. A low surrogate is present at the beginning.
  2. A high surrogate is present at the last.
  3. No low surrogate after a high surrogate.
  4. No high surrogate before a low surrogate.

> What is the purpose of this pull request?

This pull request:

-   Adds a package to check if a string is well-formed or not.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #1065 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

Screenshot of tests:

Screenshot 2024-02-26 at 10 53 57

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   ☑️  Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

File: lib/node_modules/@stdlib/assert/is-well-formed-string/package.json [hunk 1]
```diff
@@ -0,0 +1,71 @@
+{
+  "name": "@stdlib/assert/is-negative-number",
+  "version": "0.0.0",
+  "description": "Test if a value is a number having a negative value.",
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/package.json]: @gunjjoshi Could you please update the name, description, and keywords in the `package.json`?

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/primitive.js [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+* var bool = isWellFormedString( '\uDC00' );
+* // returns false
+*
+* @example
+* var bool = isWellFormedString( new String( '' ) );
+* // returns false
+*/
+function isWellFormedString( str ) {
+	return (
+		typeof str === 'string' && isWellFormed(str)
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/primitive.js]:
```suggestion
		typeof str === 'string' && isWellFormed( str )
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/README.md [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+var bool = isWellFormedString( '' );
+// returns true
+bool = isWellFormedString( new String( '' ) );
+// returns true
+bool = isWellFormedString( '\uD800\uD800' );
+// returns false
+bool = isWellFormedString( '\uDBFF' );
+// returns false
+bool = isWellFormedString( null );
+// returns false
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/README.md]:
```suggestion
var bool = isWellFormedString( '' );
// returns true

bool = isWellFormedString( new String( '' ) );
// returns true

bool = isWellFormedString( '\uD800\uD800' );
// returns false

bool = isWellFormedString( '\uDBFF' );
// returns false

bool = isWellFormedString( null );
// returns false
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/README.md [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+Tests if a `string` is a well-formed `String` object.
+
+<!-- eslint-disable no-new-wrappers -->
+
+```javascript
+var bool = isWellFormedString.isObject( '' );
+// returns false
+bool = isWellFormedString.isObject( new String( '' ) );
+// returns true
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/README.md]:
```suggestion
var bool = isWellFormedString.isObject( '' );
// returns false

bool = isWellFormedString.isObject( new String( '' ) );
// returns true
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/README.md [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+bool = isWellFormedString( '\uDBFF' );
+// returns false
+bool = isWellFormedString( '\uDBFF\uDBFF' );
+// returns false
+bool = isWellFormedString( [] );
+// returns false
+bool = isWellFormedString( '-5' );
+// returns true
+bool = isWellFormedString( null );
+// returns false
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/README.md]:
```suggestion
var isWellFormedString = require( '@stdlib/assert/is-well-formed-string' );

var bool = isWellFormedString( '' );
// returns true

bool = isWellFormedString( new String( '' ) );
// returns true

bool = isWellFormedString( '\uDBFF' );
// returns false

bool = isWellFormedString( '\uDBFF\uDBFF' );
// returns false

bool = isWellFormedString( [] );
// returns false

bool = isWellFormedString( '-5' );
// returns true

bool = isWellFormedString( null );
// returns false
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/README.md [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+<!-- Section for related `stdlib` packages. Do not manually edit this section, as it is automatically populated. -->
+
+<section class="related">
+
+* * *
+
+## See Also
+
+-   <span class="package-name">[`@stdlib/assert/is-String`][@stdlib/assert/is-String]</span><span class="delimiter">: </span><span class="description">test if a str is a String.</span>
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/README.md]: No reason to manually populate this section; we have tooling that automatically takes care of it.

```suggestion
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/README.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+<!-- /.related -->
+
+<!-- Section for all links. Make sure to keep an empty line after the `section` element and another before the `/section` close. -->
+
+<section class="links">
+
+<!-- <related-links> -->
+
+[@stdlib/assert/is-string]: https://github.com/stdlib-js/stdlib/tree/develop/lib/node_modules/%40stdlib/assert/is-string
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/README.md]:
```suggestion
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+}
+
+// The compiler throws an error if the isPrimitive method is provided an unsupported number of arguments...
+{
+	isWellFormedString.isPrimitive(); // $ExpectError
+	isWellFormedString.isPrimitive( '', '\uDBFF' ); // $ExpectError
+}
+
+
+// Attached to main export is an isPrimitive method which returns a boolean...
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts]:
```suggestion
// Attached to main export is an isObject method which returns a boolean...
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+	isWellFormedString.isPrimitive( new String( '' ) ); // $ExpectType boolean
+	isWellFormedString.isPrimitive( '' ); // $ExpectType boolean
+}
+
+// The compiler throws an error if the isPrimitive method is provided an unsupported number of arguments...
+{
+	isWellFormedString.isPrimitive(); // $ExpectError
+	isWellFormedString.isPrimitive( '', '\uDBFF' ); // $ExpectError
+}
+
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/docs/types/test.ts:48]:
```suggestion
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/examples/index.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+'use strict';
+
+var isWellFormedString = require( './../lib' );
+
+console.log( isWellFormedString( '' ) );
+// => true
+
+console.log( isWellFormedString( new String( '' ) ) );
+// => false
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/examples/index.js]: Shouldn't this be `true`? Am confused why linting didn't pick on this...

```suggestion
// => true
```

**Review by Planeshifter — CHANGES_REQUESTED** [review]
Thanks for working on this! Looks very good; left some comments that will need to be addressed before merging.

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+* @example
+* var bool = isWellFormed( '' );
+* // returns true
+*
+* @example
+* var bool = isWellFormed( new String( '\uDC00' ) );
+* // returns false
+*/
+function isWellFormed(str) {
+	for ( i = 0; i < str.length; i++) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
	for ( i = 0; i < str.length; i++ ) {
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+		// Checking if a low surrogate is present at the beginning
+		if (i === 0 && RE_UTF16_LOW_SURROGATE.test(str[i])) {
+			return false;
+		}
+
+		// Checking if a high surrogate is present at the last
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
		// Checking if a high surrogate is present at the last position
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+		if (i === str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i])) {
+			return false;
+		}
+
+		// Checking if there is no low surrogate after a high surrogate
+		if (i < str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i])) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
		if ( i < str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i]) ) {
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+			if (RE_UTF16_LOW_SURROGATE.test(str[i + 1])) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
			if ( RE_UTF16_LOW_SURROGATE.test(str[i + 1]) ) {
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+				i+=1;
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
				i += 1;
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+			}
+			else {
+				return false;
+			}
+		}
+
+		// Checking if there is no high surrogate before a low surrogate
+		if ( i>0 && RE_UTF16_LOW_SURROGATE.test(str[i])) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
		if ( i > 0 && RE_UTF16_LOW_SURROGATE.test(str[i]) ) {
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+			if (!RE_UTF16_HIGH_SURROGATE.test(str[i - 1])) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
			if ( !RE_UTF16_HIGH_SURROGATE.test(str[i - 1]) ) {
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+* limitations under the License.
+*/
+
+'use strict';
+
+// VARIABLES //
+
+var RE_UTF16_LOW_SURROGATE = /[\uDC00-\uDFFF]/;
+var RE_UTF16_HIGH_SURROGATE = /[\uD800-\uDBFF]/;
+var i = 0;
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]: We should not initialize the `i` variable outside the function scope. Just declare `var i;` in L45 before the loop.

File: lib/node_modules/@stdlib/assert/is-well-formed-string/package.json (hunk 1 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/assert/is-well-formed-string/package.json]: Sure, I've made the suggested changes.

File: lib/node_modules/@stdlib/assert/is-well-formed-string/examples/index.js (hunk 10 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/assert/is-well-formed-string/examples/index.js]: Yes, it should be true.

Comment (Planeshifter) [reply lib/node_modules/@stdlib/assert/is-well-formed-string/examples/index.js]: Thanks for confirming!

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js (hunk 18 above)
Comment (gunjjoshi) [reply lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]: I assumed that all variables should be declared in the variables section. 
But yes, it is outside the scope of the function, and we should avoid using global variables.
I've made the suggested change.

**Comment by gunjjoshi** [comment]
@Planeshifter 
There is some error in linting the examples. I have written the correct outputs, but it is expecting something else, and thus giving errors. I checked the implementation of the functions, it is indeed correct. 

Screenshot 2024-02-27 at 10 36 46

**Comment by Planeshifter** [comment]
@gunjjoshi Thanks for checking! I believe the issue is that the `instanceof String` check doesn't work across realms when there are multiple `String` constructors. To lint examples and their expected outputs, we spawn a separate process via `vm2` that will have it's own `String` constructor. Here is how we make the object check in the `assert/is-string` package https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/assert/is-string/lib/object.js more robust to have it work across realms.
Would you mind updating this package accordingly?

**Comment by gunjjoshi** [comment]
> @gunjjoshi Thanks for checking! I believe the issue is that the `instanceof String` check doesn't work across realms when there are multiple `String` constructors. To lint examples and their expected outputs, we spawn a separate process via `vm2` that will have it's own `String` constructor. Here is how we make the object check in the `assert/is-string` package https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/assert/is-string/lib/object.js more robust to have it work across realms. Would you mind updating this package accordingly?

Sure, I'll update this package.

**Comment by gunjjoshi** [comment]
Great ! It got passed by the CI pipeline.
So the string object checking was creating the problem.

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+*
+* @example
+* var bool = isWellFormed( new String( '\uDC00' ) );
+* // returns false
+*/
+function isWellFormed(str) {
+	var i;
+	for ( i = 0; i < str.length; i++ ) {
+		// Checking if a low surrogate is present at the beginning
+		if (i === 0 && RE_UTF16_LOW_SURROGATE.test(str[i])) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
		if ( i === 0 && RE_UTF16_LOW_SURROGATE.test(str[i]) ) {
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js [hunk 20]
```diff
@@ -0,0 +1,79 @@
...
+			return false;
+		}
+
+		// Checking if a high surrogate is present at the last position
+		if (i === str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i])) {
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/iswellformed.js]:
```suggestion
		if ( i === str.length - 1 && RE_UTF16_HIGH_SURROGATE.test(str[i]) ) {
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/object.js [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+* var bool = isWellFormedString( '\uDC00' );
+* // returns false
+*
+* @example
+* var bool = isWellFormedString( new String( '' ) );
+* // returns true
+*/
+function isWellFormedString( str ) {
+	return (
+		isString.isObject(str) &&
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/object.js]:
```suggestion
		isString.isObject( str ) &&
```

File: lib/node_modules/@stdlib/assert/is-well-formed-string/lib/object.js [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+		isWellFormed(str.valueOf())
```
Comment (Planeshifter) [inline lib/node_modules/@stdlib/assert/is-well-formed-string/lib/object.js]:
```suggestion
		isWellFormed( str.valueOf() )
```

**Review by Planeshifter — APPROVED** [review]

**Review by Planeshifter — APPROVED** [review]

**Comment by Planeshifter** [comment]
@gunjjoshi Will merge after CI clears; thanks again for your contribution!

**Pull request merged**